python/pyarrow/_dataset.pyx

```diff
@@ -1592,6 +1593,145 @@ cdef class HivePartitioning(Partitioning):
             res.append(pyarrow_wrap_array(arr))
         return res
 
+cdef class FilenamePartitioning(Partitioning):
```

Comment:
We should add some python tests which cover this logic.  Hopefully we can adapt some of the existing partitioning tests pretty easily.

Reply:
Added test in Python for FilenamePartitioning